### PR TITLE
HPCC-28545 Use thread-local context for better logging control

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Regression suite and examining logs

### CODE REVIEW

File: system/jlib/jlog.cpp [hunk 1]
```diff
@@ -2626,11 +2629,9 @@ class DummyLogCtx : implements IContextLogger
     virtual void Link() const {}
     virtual bool Release() const { return false; }
 
-    virtual void CTXLOGva(const char *format, va_list args) const __attribute__((format(printf,2,0)))
+    virtual void CTXLOGva(const LogMsgCategory & cat, const LogMsgJobInfo & job, LogMsgCode code, const char *format, va_list args) const override  __attribute__((format(printf,5,0)))
```
Comment (ghalliday) [inline system/jlib/jlog.cpp:2632]: trivial: trailing space on end of line

File: system/jlib/jlog.cpp [hunk 2]
```diff
@@ -3185,3 +3191,70 @@ TraceFlags loadTraceFlags(const IPropertyTree *ptree, const std::initializer_lis
     }
     return dft;
 }
+
+void ctxlog_report(const LogMsgCategory & cat, const char * format, ...)
```
Comment (ghalliday) [inline system/jlib/jlog.cpp:3195]: Normally we use camel case any reason for the a_b style?

File: system/jlib/jlog.cpp [hunk 3]
```diff
@@ -3185,3 +3191,70 @@ TraceFlags loadTraceFlags(const IPropertyTree *ptree, const std::initializer_lis
     }
     return dft;
 }
+
+void ctxlog_report(const LogMsgCategory & cat, const char * format, ...) 
+{
+    va_list args;
+    va_start(args, format);
+    ctxlog_report_va(cat, unknownJob, NoLogMsgCode, format, args);
```
Comment (ghalliday) [inline system/jlib/jlog.cpp:3199]: defaultJobInfo rather than unknownJob?

**Review by ghalliday — COMMENTED** [review]
@richardkchapman a couple of minor comments

File: system/jlib/jlog.cpp (hunk 1 above)
Comment (richardkchapman) [reply system/jlib/jlog.cpp:2632]: Will fix

File: system/jlib/jlog.cpp (hunk 3 above)
Comment (richardkchapman) [reply system/jlib/jlog.cpp:3199]: unknownJob is translated to defaultJobInfo at a lower level. I used unknownJob for consistency with other places, but I can change if you want

File: system/jlib/jlog.cpp (hunk 2 above)
Comment (ghalliday) [reply system/jlib/jlog.cpp:3195]: ctxlogReportVa?  Just looks out of place.

**Comment by richardkchapman** [comment]
@ghalliday Please re-review

**Review by ghalliday — APPROVED** [review]
@richardkchapman please squash

**Comment by richardkchapman** [comment]
@ghalliday Squashed

**Pull request merged**